### ISSUE TITLE
Loot percentages do not seem right?

### ISSUE THREAD

**Issue by Yiaro222** [DavidS-Repo/SkyGridx#10]
**Describe the Bug**
I created a test environment to look through what the chestloot would look like with a certain set of chances and loot. I changed the chest settings for the chest that, by default, was meant for the end. I also made it easier to find chests by changing the world generation to spawn with more chests. The weight of the chestsettings does not seem right.

**To Reproduce**
Steps to reproduce the behavior & Configuration files:
Change these specific files to have the text under them to recreate the behavior.

SkyGrid/SkygridBlocks/world_the_end.txt: 
CHEST:0.01
OBSIDIAN: 1

SkyGrid/SkygridBlocks/ChestSettings.yml:
 Chest4:
   Items: [ELYTRA:5:1, ENDER_PEARL:5:3, SHULKER_SHELL:5:1, END_CRYSTAL:5:1]
   Biomes: [THE_END, END_BARRENS, END_HIGHLANDS, END_MIDLANDS, SMALL_END_ISLANDS]

**Expected Behavior**
With the values 5 and 1 behind the material name i expected a 5% chance to get 1 of the material, instead i got more of an 85% chance to get 1 of the material and i do not understand why. I have tried to understand the file: [SkyGridx](https://github.com/DavidS-Repo/SkyGridx/tree/main)/[SkyGridPlugin](https://github.com/DavidS-Repo/SkyGridx/tree/main/SkyGridPlugin)/[main](https://github.com/DavidS-Repo/SkyGridx/tree/main/SkyGridPlugin/main)/[CustomChest.java](https://github.com/DavidS-Repo/SkyGridx/blob/main/SkyGridPlugin/main/CustomChest.java)

but i could not figure out how exactly the weight is used in the code. I am probably doing something wrong but i dont know what.

**Screenshots/Logs (Optional)**
The average chest with the chestsettings as shown above:
![afbeelding](https://github.com/user-attachments/assets/2f3ddb5c-6fa7-4720-b11f-031ed7a6c627)

**Server Information (Please complete the following):**
- **Minecraft Version**: 1.21,4
- **Server Type**: Spigot
- **Other Plugins Installed (Optional)**: https://www.spigotmc.org/resources/set-home.32748/, https://www.spigotmc.org/resources/disablecommands-1-13-x-1-20-x.110823/

**Additional Context**
I am not the best at this, i am trying to learn a lot of things surrounding all of this.
I hope to have given enough context in this way.

Oh and btw: Happy new years (maybe you're reading this in 2025)

**Comment by DavidS-Repo** [comment]
**Hi!**

It looks like the loot distribution you’re seeing is actually the intended behavior of the plugin.

### Relative Weights & “Hidden” Probabilities
- Each item in your `Items` list has a `weight` value.
- The plugin sums **all** these weights. Then, to choose an item for each empty slot, it picks one based on how large its weight is relative to the total.
- If you have four items each with weight 5 (like ELYTRA, ENDER_PEARL, SHULKER_SHELL, END_CRYSTAL), you get a total of 20.  
  - Each item then has a per-slot probability of `5/20 = 25%`.
- The code automatically normalizes the weights to distribute items by probability.

### How to Reduce the Overall Item Chance
- Because this process happens **per slot**, if each slot has a 25% chance to get ELYTRA, you’ll likely see multiple ELYTRAs across many empty slots (at least until the item hits its `maxAmount`).
- To get a **lower** total chance, say you only want around 5% across the **entire 27-slot chest**, you need to reduce that per-slot probability significantly.

### Using a Dummy “AIR” Item
- A common approach is to add a “dummy” or filler item with a high weight, often `AIR` (this results in more empty slots), so that the total weight is much larger and the chance for real items becomes smaller.
- For example:

```yml
ChestSettings:
  Chest1:
    Items:
      - AIR:80:64
      - ELYTRA:5:1
      - ENDER_PEARL:5:3
      - SHULKER_SHELL:5:1
      - END_CRYSTAL:5:1
```

- Now, your total weight is `80 + 5 + 5 + 5 + 5 = 100`.  
  - Each real item is `5/100 = 5% per slot`.  
- But remember, if you’re filling **all 27 slots**, multiple slots can still roll that 5% chance, which might increase the probability of seeing at least one ELYTRA to more than 5%. 
- If that’s **still** too high, you can raise the `AIR` weight more, or lower the other items’ weights even further.

### Distribution vs. Exact Percent Goals
- If you want precisely 5% chance or close to it for the **whole chest** (not per slot), you’d have to do some math for "1 - (chance of no item in each slot)^(number of slots)” or keep increasing your `AIR` weight until it feels right.
- Either way, the plugin’s method is to choose an item (or “AIR”) **per empty slot** based on these relative weights. It’s not a bug; it’s how the loot table was designed to work.

### TL;DR
- The plugin normalizes weights to choose items by probability per slot.  
- Adding a dummy `AIR` item with a large weight can dilute the real items’ chances and get you closer to your desired distribution.  Or you can just add less desirable items like `Chorus Fruit` or something like that to make up the dummy items that will take up the majority of the odds.
- If you’re aiming for a certain overall **chest-wide** percentage (like 5%), be aware that multiple slots each rolling 5% can yield higher than 5% across the entire chest, so you’ll likely need to adjust weights further.

I probably did a bad job explaining this in the files, and it's also quite outdated so I will update the descriptions in the next update.

**Hope this helps clarify everything!** Let me know if you have any more questions or need more details.

**Comment by Yiaro222** [comment]
Thank you so much for your extensive response!

All of what you explained answered every single question i had about the chest loot generation. So thank you very much for that! 

Reading the CustomChest.java, i had figured out it added all of its weights but i thought i was missing something. But all of that is answered now.
Sorry for putting it as a bug, i wasnt sure how to create a ticket in any other form. 

It now makes sense:
- why a 100% chance stone block always appeared in the same slot (i had gotten confused why even after this 100% stone block the propabilities were still so high, after looking through the code)
- the chance system works
- that the chance system is chance per slot

I am trying to recreate sethblings loot chances but with all of the newer items aswell. With what you told me i can continue this.


Everything just makes sense now. Thank you so much for your response, how quickly you responded, the extensiveness of the response, the examples AND solutions for my personal problem!
I will take a look at the chestloot system as soon as i have the time and give you an update. For now i do not have any further questions. 
Again, thank you.

**Comment by DavidS-Repo** [comment]
Glad everything makes sense now! If you run into anything else or have more questions down the line, feel free to reopen the issue. Happy to help, good luck with the loot system!
